### ISSUE TITLE
OSS: RBAC fail to authorize user coming from a trusted cluster

### ISSUE THREAD

**Issue by apestel** [gravitational/teleport#1674]
**What happened**:
Remote user from main remote cluster with root access can't connect to local node in local cluster.
RBAC implementation coming with OSS version block access with ImplicitDefaultRole missing Logins and NodeLabels "*":"*".

**What you expected to happen**:
RBAC should accept incoming connection request based on the ImplicitDefaultRole

**How to reproduce it (as minimally and precisely as possible)**:
- remote auth + proxy (2.4.1)
- local auth + proxy + node (2.4.1)
Try to connect from remote proxy to local node with root account.

**Environment**:
- Teleport version (use `teleport version`): 2.4.1
- Tsh version (use `tsh version`): web
- OS (e.g. from /etc/os-release): Debian/jessie

**Relevant Debug Logs If Applicable**
WARN             [RBAC] Denied access to server: access to node nodetest is denied to role(s): [role default-implicit-role denied access: allow rules did not match; match(namespace=true, label=false, login=false)] services/role.go:1487
ERRO [NODE]      Permission denied: user myuser@main is not authorized to login as root@ca-test: access to node nodetest is denied to role(s): [role default-implicit-role denied access: allow rules did not match; match(namespace=true, label=false, login=false)] fingerprint:ssh-rsa-cert-v01@openssh.com SHA256:8pbFsMUpkhHjC5DCtKfHeJ/AvH5ksQ2Qp/1A7bYQJ+o local:<local_ip>:3022 remote:<remote_ip>:58361 user:root srv/authhandlers.go:216
WARN [NODE]      failed login attempt events.EventFields{"user":"myuser", "success":false, "error":"user myuser@main is not authorized to login as root@ca-test: access to node nodetest is denied to role(s): [role default-implicit-role denied access: allow rules did not match; match(namespace=true, label=false, login=false)]"} fingerprint:ssh-rsa-cert-v01@openssh.com SHA256:8pbFsMUpkhHjC5DCtKfHeJ/AvH5ksQ2Qp/1A7bYQJ+o local:<local_ip>:3022 remote:<remote_ip>:58361 user:root srv/authhandlers.go:166

below patch that fix this issue:

```
diff --git a/lib/services/role.go b/lib/services/role.go
index 0eab977b..87df9c64 100644
--- a/lib/services/role.go
+++ b/lib/services/role.go
@@ -124,6 +124,8 @@ func NewImplicitRole() Role {
                        Allow: RoleConditions{
                                Namespaces: []string{defaults.Namespace},
                                Rules:      CopyRulesSlice(DefaultImplicitRules),
+                               Logins:     []string{"root", "admin"},
+                               NodeLabels: map[string]string{"*": "*"},
                        },
                },
        }

```

**Comment by russjones** [comment]
@apestel I'm not quite sure I understand the issue.

Using the example in Teleport documentation [1], have you established a trust relationship between "main" and "east"?

If so, this relationship means that users in the "main" cluster can connect to nodes in the "east" cluster, not the other way around (meaning users in the "east" cluster can not connect to nodes in "main").

[1] https://gravitational.com/teleport/docs/2.4/admin-guide/#trusted-clusters

**Comment by apestel** [comment]
There is a trust relationship between both.
The connection come from main to east.

Without the provided patch it doesn't work.
With it, it works.

**Comment by russjones** [comment]
@apestel Can you show me how you are creating users on both clusters and how you are establishing the trust relationship? Then I can try and reproduce the issue.

**Comment by aashley** [comment]
I've started seeing this as well. Nodes that where authenticating correctly with 2.4.0 are now failing with this error after upgrading the remote auth node to 2.4.1. This only affects nodes that added a trust cluster after upgrading to 2.3.0, those that initially where setup with the 2.2 and earlier method are still working.

User in teleport cloud cluster
```
root@teleport:/var/lib# tctl get user/adama
kind: user
metadata:
  name: adama
spec:
  created_by:
    time: 0001-01-01T00:00:00Z
    user:
      name: ""
  expires: 0001-01-01T00:00:00Z
  roles:
  - admin
  status:
    is_locked: false
    lock_expires: 0001-01-01T00:00:00Z
    locked_time: 0001-01-01T00:00:00Z
  traits:
    logins:
    - root
    - swift
version: v2
```
teleport remote node cluster
```
root@od-server-00202:/var/log# tctl get clusters
kind: trusted_cluster
metadata:
  name: cloud
spec:
  enabled: true
  role_map:
  - local:
    - admin
    remote: admin
  token: xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
  tunnel_addr: XX.XX.XXX.XX:3024
  web_proxy_addr: XX.XX.XXX.XX:3080
version: v2
```

teleport -d from remote node during auth attempt
```
DEBU [PROXY:AGE] ping -> XX.XX.XXX.XX:3024 target:XX.XX.XXX.XX:3024 reversetunnel/agent.go:560
DEBU [PROXY:AGE] access point request: teleport-access-point target:XX.XX.XXX.XX:3024 reversetunnel/agent.go:578
DEBU [PROXY:AGE] proxyAccessPoint target:XX.XX.XXX.XX:3024 reversetunnel/agent.go:307
DEBU [CLIENT:TU] dialing tcp accesspoint:0 purpose:Proxy auth/tun.go:925
DEBU [CLIENT:TU] auth servers: [{127.0.0.1:3025 tcp } {0.0.0.0:3025 tcp } {10.222.117.66:3025 tcp }] purpose:Proxy auth/tun.go:1047
DEBU [CLIENT:TU] dialing 127.0.0.1:3025, attempt 1 purpose:Proxy auth/tun.go:1071
INFO             [AUTH] keyAuth: 127.0.0.1:39672->127.0.0.1:3025, user=9f8c49f7-749a-41da-96b8-45da23f667b1.od-server-00202 auth/tun.go:423
DEBU [SSH:AUTH]  incoming connection 127.0.0.1:39672 -> 127.0.0.1:3025 vesion: SSH-2.0-Go sshutils/server.go:331
DEBU             [AUTH] new channel request for direct-tcpip from 127.0.0.1:39672 auth/tun.go:175
INFO [PRESENCE]  GetServers(1) in 1.523619ms local/presence.go:200
DEBU [PROXY:AGE] transport request: teleport-transport target:XX.XX.XXX.XX:3024 reversetunnel/agent.go:590
DEBU [PROXY:AGE] proxyTransport target:XX.XX.XXX.XX:3024 reversetunnel/agent.go:348
DEBU [PROXY:AGE] Received out-of-band proxy transport request: [127.0.0.1:3022] target:XX.XX.XXX.XX:3024 reversetunnel/agent.go:400
DEBU [PROXY:AGE] successfully dialed to 127.0.0.1:3022, start proxying target:XX.XX.XXX.XX:3024 reversetunnel/agent.go:427
DEBU [PROXY:AGE] ping -> XX.XX.XXX.XX:3024 target:XX.XX.XXX.XX:3024 reversetunnel/agent.go:560
DEBU [NODE]      conn(127.0.0.1:51086->127.0.0.1:3022, user=root) auth attempt fingerprint:ssh-rsa-cert-v01@openssh.com SHA256:+9iXw6Wzk9ruk4Ke7I2aENMTISBwMRyhDgMhjYhIyGY local:127.0.0.1:3022 remote:127.0.0.1:51086 user:root srv/authhandlers.go:140
DEBU [NODE]      conn(127.0.0.1:51086->127.0.0.1:3022, user=root) auth attempt with key ssh-rsa-cert-v01@openssh.com SHA256:+9iXw6Wzk9ruk4Ke7I2aENMTISBwMRyhDgMhjYhIyGY, &ssh.Certificate{Nonce:[]uint8{0x19, 0xb0, 0x92, 0x2c, 0x4c, 0x94, 0x2d, 0x3e, 0xef, 0x4e, 0x22, 0x2f, 0x6a, 0x46, 0x47, 0x6d, 0x16, 0x3c, 0x6, 0xa5, 0x78, 0xf5, 0x82, 0x8d, 0x86, 0x6b, 0xeb, 0x9e, 0xc6, 0x5f, 0xd3, 0xb1}, Key:(*ssh.rsaPublicKey)(0xc420138860), Serial:0x0, CertType:0x1, KeyId:"adama", ValidPrincipals:[]string{"root", "swift"}, ValidAfter:0x0, ValidBefore:0x5a8aecdd, Permissions:ssh.Permissions{CriticalOptions:map[string]string{}, Extensions:map[string]string{"permit-agent-forwarding":"", "permit-port-forwarding":"", "permit-pty":""}}, Reserved:[]uint8{}, SignatureKey:(*ssh.rsaPublicKey)(0xc4201388a0), Signature:(*ssh.Signature)(0xc420140ae0)} fingerprint:ssh-rsa-cert-v01@openssh.com SHA256:+9iXw6Wzk9ruk4Ke7I2aENMTISBwMRyhDgMhjYhIyGY local:127.0.0.1:3022 remote:127.0.0.1:51086 user:root srv/authhandlers.go:143
DEBU [NODE]      Successfully authenticated fingerprint:ssh-rsa-cert-v01@openssh.com SHA256:+9iXw6Wzk9ruk4Ke7I2aENMTISBwMRyhDgMhjYhIyGY local:127.0.0.1:3022 remote:127.0.0.1:51086 user:root srv/authhandlers.go:180
DEBU [NODE]      Checking permissions for (adama,root) to login to node with RBAC checks. fingerprint:ssh-rsa-cert-v01@openssh.com SHA256:+9iXw6Wzk9ruk4Ke7I2aENMTISBwMRyhDgMhjYhIyGY local:127.0.0.1:3022 remote:127.0.0.1:51086 user:root srv/authhandlers.go:308
WARN             [RBAC] Denied access to server: access to node od-server-00202 is denied to role(s): [role default-implicit-role denied access: allow rules did not match; match(namespace=true, label=false, login=false)] services/role.go:1487
ERRO [NODE]      Permission denied: user adama@cloud is not authorized to login as root@od-server-00202: access to node od-server-00202 is denied to role(s): [role default-implicit-role denied access: allow rules did not match; match(namespace=true, label=false, login=false)] fingerprint:ssh-rsa-cert-v01@openssh.com SHA256:+9iXw6Wzk9ruk4Ke7I2aENMTISBwMRyhDgMhjYhIyGY local:127.0.0.1:3022 remote:127.0.0.1:51086 user:root srv/authhandlers.go:216
WARN [NODE]      failed login attempt events.EventFields{"user":"adama", "success":false, "error":"user adama@cloud is not authorized to login as root@od-server-00202: access to node od-server-00202 is denied to role(s): [role default-implicit-role denied access: allow rules did not match; match(namespace=true, label=false, login=false)]"} fingerprint:ssh-rsa-cert-v01@openssh.com SHA256:+9iXw6Wzk9ruk4Ke7I2aENMTISBwMRyhDgMhjYhIyGY local:127.0.0.1:3022 remote:127.0.0.1:51086 user:root srv/authhandlers.go:166
DEBU [AUDITLOG]  EmitAuditEvent(auth: map[success:false user:adama error:user adama@cloud is not authorized to login as root@od-server-00202: access to node od-server-00202 is denied to role(s): [role default-implicit-role denied access: allow rules did not match; match(namespace=true, label=false, login=false)]]) events/auditlog.go:301
DEBU [NODE]      conn(127.0.0.1:51086->127.0.0.1:3022, user=root) auth attempt fingerprint:ssh-rsa SHA256:TmUM9fiB8jCQhFKRlDaGnOjmG9f5CoGoP4o7EeKCPSo local:127.0.0.1:3022 remote:127.0.0.1:51086 user:root srv/authhandlers.go:140
DEBU [NODE]      conn(127.0.0.1:51086->127.0.0.1:3022, user=root) auth attempt with key ssh-rsa SHA256:TmUM9fiB8jCQhFKRlDaGnOjmG9f5CoGoP4o7EeKCPSo, (*ssh.Certificate)(nil) fingerprint:ssh-rsa SHA256:TmUM9fiB8jCQhFKRlDaGnOjmG9f5CoGoP4o7EeKCPSo local:127.0.0.1:3022 remote:127.0.0.1:51086 user:root srv/authhandlers.go:143
DEBU [NODE]      auth attempt, unsupported key type fingerprint:ssh-rsa SHA256:TmUM9fiB8jCQhFKRlDaGnOjmG9f5CoGoP4o7EeKCPSo local:127.0.0.1:3022 remote:127.0.0.1:51086 user:root srv/authhandlers.go:146
DEBU [PROXY:AGE] ping -> XX.XX.XXX.XX:3024 target:XX.XX.XXX.XX:3024 reversetunnel/agent.go:560
DEBU [PROXY:AGE] ping -> XX.XX.XXX.XX:3024 target:XX.XX.XXX.XX:3024 reversetunnel/agent.go:560
```

**Comment by russjones** [comment]
@aashley @apestel I'm having some problems reproducing this. This is what I did:

1. Create a fresh Teleport 2.4.0 clusters called `main`.
1. Create a fresh Teleport 2.4.0 cluster called `east`.
1. Establish a Trusted Cluster relationship where `east` is the remote cluster.
1. Upgrade the `east` cluster to Teleport 2.4.1.
1. Connect from `main` into a node in `east`.

I was able to do the above successfully.

Can you provide the steps you took so I can reproduce this issue?

**Comment by aashley** [comment]
@russjones I had a similar in our test cluster before deploying to production. It was only upon upgrading the production systems that where installed with 2.1 and have been upgraded since that I started getting the failures.

Upon rolling back to 2.4.0 in the cloud cluster authentication started working.

We had a similar issue in upgrading to 2.4.0 where we had to recreate the user accounts to get them with the correct roles for the trust accounts to work, but from what I can see there is no such change between 2.4.0 and 2.4.1.

- Upgrade 2.4.0 cluster to 2.4.1 called cloud, created with 2.1.0
- Upgrade 2.4.0 cluster to 2.4.1 called node1, created with 2.1.0
- Upgrade 2.4.0 cluster to 2.4.1 called node2, created with 2.4.0
- Keep 2.4.0 cluster called node3, created with 2.1.0
- Keep 2.4.0 cluster called node4, created with 2.4.0
- Connect from cloud to node in node1: works
- Connect from cloud to node in node2: fails
- Connect from cloud to node in node3: works
- Connect from cloud to node in node4: fails
- Revert cloud to 2.4.0
- Connect from cloud to node in node1: works
- Connect from cloud to node in node2: fails
- Connect from cloud to node in node3: works
- Connect from cloud to node in node4: works
- Revert node1 & node 2 to 2.4.0
- All connection options work.

**Comment by russjones** [comment]
@aashley @apestel The root cause of this issue was a block of code was not backported from `master` to our 2.4 release branch. I fixed that in #1707, it will be released with 2.4.2. If this is still an issue in either 2.5.0 or 2.4.2 feel free to open another issue.
